### PR TITLE
Get rid of reflection blocked types

### LINKED ISSUE #72570: Investigate getting rid of reflection blocking
NativeAOT has a significant amount of the runtime written in C#. Keeping reflection metadata for it results in a lot of garbage being generated by the compiler since nobody is ever going to reflect on these things (and if they do, preventing that sounds like a good thing). We have a concept of reflection blocking for that that we inherited from .NET Native.

But now that IlcTrimMetadata=true is the default, there's little size incentive in keeping this - we should measure the impact. If it's not significant, we should stop doing reflection blocking.

We currently disable reflection blocking in the libs test because they do reflect on CoreLib's implementation details to test them. But it also means we're not running the tests in the configuration that the users will run. It could mask issues such as private reflection being done by the libraries themselves (not just by library tests).

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #72570.

Still need to delete workarounds that make things public in corelib but maybe this diff is large enough already?

We were gradually getting less and less from reflection blocking:

* We stopped blocking things outside corelib (.NET Native blocked all of BCL; we don't).
* We trim reflection metadata and that allows us to have method bodies without metadata.

With this, we should be able to get type definition-level reflection metadata for any MethodTable there is in the system.

This is a ~30 kB size regression for BasicMinimalApis. It is pretty much a wash for Hello World, because all the BCL cruft to handle reflection blocked types costs as much as the benefit of blocking.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by MichalStrehovsky** [comment]
Unfortunately, the test failures are real. There is one other kind of `MethodTable`s that take the "reflection blocked" route at runtime: `MethodTable`s that are not actually user visible.

For example `someType == typeof(NeverAllocated)` would not create a constructed `MethodTable` and as a result would not get metadata. But we now need the metadata to create the `RuntimeType` instance since the reflection blocked path was deleted.

I'm fixing it by giving necessary MethodTables reflection metadata. It means they're now user visible and users can browse to them with e.g. `Type.GetType`. This should only be a problem for trim-unsafe code. I think the only problem with that would be `GetUninitializedObject` so I'm adding a new check for that to prevent these from landing on the GC heap (which would be a lot more unsafe than trimming-unsafe).

**Comment by MichalStrehovsky** [comment]
Okay, so on Linux the regression on Hello world is much worse - ~1,569 kB to 1,821 kB. Need to figure that out first 😭.

**Comment by MichalStrehovsky** [comment]
The Linux regression was the same problem as https://github.com/dotnet/runtime/pull/84156#issuecomment-1491334631 with the SpanAction thing. @jkotas  could you have a look at that commit? I made a small functional change in it because it looks like we were passing a wrong size of the buffer so I fixed that.

File: src/libraries/System.Private.CoreLib/src/System/Text/Latin1Encoding.cs [hunk 1]
```diff
@@ -586,14 +583,13 @@ public override unsafe string GetString(byte[] bytes, int index, int count)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.bytes, ExceptionResource.ArgumentOutOfRange_IndexCountBuffer);
             }
 
-            return string.Create(count, (encoding: this, bytes, index), static (chars, args) =>
+            string result = string.FastAllocateString(bytes.Length);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Text/Latin1Encoding.cs:586]:
```suggestion
            string result = string.FastAllocateString(count);
```
I do not think that `bytes.Length` is correct. This method does not convert the whole array. Let's see whether we have tests to catch this bug.

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Text/Latin1Encoding.cs]: Yup, we do:

```
Test System.Text.Tests.Latin1EncodingDecode.Decode has failed. Check the Test tab for details.
```

File: src/libraries/System.Private.CoreLib/src/System/Text/Latin1Encoding.cs [hunk 2]
```diff
@@ -586,14 +583,13 @@ public override unsafe string GetString(byte[] bytes, int index, int count)
...
+            fixed (byte* pBytes = bytes)
+            fixed (char* pChars = result)
             {
-                fixed (byte* pBytes = args.bytes)
-                fixed (char* pChars = chars)
-                {
-                    args.encoding.GetCharsCommon(pBytes + args.index, chars.Length, pChars, chars.Length);
-                }
-            });
+                GetCharsCommon(pBytes + index, bytes.Length - index, pChars, result.Length);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Text/Latin1Encoding.cs:590]:
```suggestion
                GetCharsCommon(pBytes + index, count, pChars, count);
```
Dtto

**Comment by agocke** [comment]
Nice! Thanks for all the hard work here @MichalStrehovsky and @jkotas, much appreciated

**Pull request merged**